### PR TITLE
JAMES-XXXX Add Filter versioning support

### PR DESCRIPTION
**Pull request by quantranhong1999**
This ensures the clients is aware of filter changes

### CODE REVIEW

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rules.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        return Objects.hash(rules, version);
+    }
+
+    @Override
+    public String toString() {
+        return "Rules{" +
+            "rules=" + rules +
+            ", version=" + version +
+            '}';
+    }
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rules.java:63]: This toString is not compliant.Use guava's `ToStringHelper`.

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Version.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    public int hashCode() {
+        return Objects.hash(version);
+    }
+
+    @Override
+    public String toString() {
+        return "Version{" +
+            "version=" + version +
+            '}';
+    }
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Version.java:52]: Idem

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Version.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.api.filtering;
+
+import java.util.Objects;
+
+public class Version {
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Version.java:24]: Can we test this POJO against EqualsVerifier ?

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rules.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            Rules that = (Rules) o;
+            return Objects.equals(this.rules, that.rules)
+                && Objects.equals(this.version, that.version);
+        }
+        return false;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(rules, version);
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rules.java:54]: Can we test this POJO against EqualsVerifier ?

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/impl/FilteringAggregate.java [hunk 5]
```diff
@@ -22,14 +22,19 @@
...
 import org.apache.james.jmap.api.filtering.Rule;
+import org.apache.james.jmap.api.filtering.Rules;
+import org.apache.james.jmap.api.filtering.Version;
 
 import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableList;
 
 public class FilteringAggregate {
 
+    public static final int NO_HISTORY = -1;
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/impl/FilteringAggregate.java:36]: Where is this used?

**Comment by chibenwa** [comment]
Please open this pull request against apache/master

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/impl/FilteringAggregate.java (hunk 5 above)
Comment (quantranhong1999) [reply server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/impl/FilteringAggregate.java:36]: no where, i will remove it

**Comment by chibenwa** [comment]
BTW can you create an Apache James ticket or this?

**Comment by quantranhong1999** [comment]
> BTW can you create an Apache James ticket or this?

Yes. I will do it

**Comment by quantranhong1999** [comment]
Moved to #https://github.com/apache/james-project/pull/341

**Pull request closed without merging**